checks/cs50/2017/x/finance/checks.py

```diff
@@ -1,13 +1,305 @@
+            raise Error("Expected quote success, not failure.")
+        elif "Apology" not in self.driver.title and not outcome:
+            raise Error("Expected quote failure, not success.")
+        elif outcome and value:
+            # check the output against the given value
+            if str(value) not in self.driver.find_element_by_tag_name("body").text:
+                raise Error("Expected quote not found.")
+
+    def transaction(self, transaction, symbol, shares, outcome, value=None, remaining=None):
+        """Checks that purchasing shares results in given outcome."""
+
+        # go to buy page
+        self.driver.get("{}/{}".format(Finance.BASE_PATH, transaction))
+        fields = self.driver.find_elements_by_tag_name("input")
+
+        # fill in fields
+        fields_filled = 0
+        for field in fields:
+            name = field.get_attribute("name")
+            if name == "symbol":
+                fields_filled += 1
+                field.send_keys(symbol)
+            elif name == "shares":
+                fields_filled += 1
+                field.send_keys(shares)
+        if fields_filled != 2:
+            raise Error("Could not fill in symbol and/or shares.")
+
+        # search for buy button
+        buttons = self.driver.find_elements_by_tag_name("button")
+        buy_button = None
+        for button in buttons:
+            if button.text == transaction.capitalize():
+                buy_button = button
+        if buy_button == None:
+            raise Error("Could not find {} button.".format(transaction))
+        buy_button.click()
+
+        if "Apology" in self.driver.title and outcome:
+            raise Error("Expected {} success, not failure.".format(transaction))
+        elif "Apology" not in self.driver.title and not outcome:
+            raise Error("Expected {} failure, not success.".format(transaction))
+
+        # go to index and check balance
+        if outcome and (value or remaining):
+            self.driver.get(Finance.BASE_PATH)
+            if value and not re.search(value, self.driver.page_source):
+                raise Error("Expected to see value {}.".format(value))
+            if remaining and not re.search(remaining, self.driver.page_source):
+                raise Error("Expected to see cash of value {}.".format(remaining))
+
+
     @check()
     def exists(self):
-        """application.c exists."""
-        super(Finance, self).exists("application.py")
+        """application.py exists."""
+        super().exists("application.py")
+
+    @check("exists")
+    def startup(self):
+        """application starts up."""
+        self.override_lookup()
+        self.server_up()
+        res = requests.get(Finance.BASE_PATH)
+        if res.status_code != 200:
+            raise Error((res.status_code, 200))
+        self.server_down()
+
+    @check("startup")
+    def can_register(self):
+        """user can register for account."""
+        self.server_up()
```

Comment:
Hm, should `server_up` and `server_down` really be the responsibility of each check? Should/could `startup` not do that? And maybe we could change `check` to take two arguments, the second of which has a default value of `None` but, if present, represents a function that should be called after this check is done, for cleanup?